File: C++/convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
   
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return buildTree(nums, 0, nums.size() - 1);
    }
    
    TreeNode* buildTree(vector<int>& nums, int i, int j) {
        if (i > j)
            return nullptr;
        int mid = (i + j) / 2;
        TreeNode* node = new TreeNode(nums[mid]);
        node->left = buildTree(nums, i, mid - 1);
        node->right = buildTree(nums, mid + 1, j);
        return node;
    }
};

File: Python3/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        return self.build(nums, 0, len(nums) - 1)
    
    def build(self, nums: List[int], lo: int, hi: int) -> Optional[TreeNode]:
        if lo <= hi:
            mid = (lo + hi) // 2
            left = self.build(nums, lo, mid-1)
            right = self.build(nums, mid+1, hi)
            node = TreeNode(nums[mid], left, right)
            return node
        return None
